pennylane/io/qualtran_io.py

```diff
@@ -418,3 +564,326 @@ def compute_matrix(*params, **hyperparams):  # pylint: disable=no-self-argument
             raise MatrixUndefinedError
 
         return matrix
+
+
+def _split_qubits(registers, qubits):  # pylint: disable=redefined-outer-name
+    """Function from the Qualtran-Cirq interop module that splits the flat list of qubits into
+    a dictionary of appropriately shaped qubit arrays."""
+
+    qubit_regs = {}
+    base = 0
+    for reg in registers:
+        qubit_regs[reg.name] = np.array(qubits[base : base + reg.total_bits()]).reshape(
+            reg.shape + (reg.bitsize,)
+        )
+        base += reg.total_bits()
+    return qubit_regs
+
+
+def _ensure_in_reg_exists(
+    bb: "qt.BloqBuilder",
+    in_reg: "_QReg",
+    qreg_to_qvar: Dict["_QReg", "qt.Soquet"],
+) -> None:
+    """Modified function from the Qualtran-Cirq interop module that takes care of qubit allocations,
+    split and joins to ensure `qreg_to_qvar[in_reg]` exists."""
+    from qualtran.cirq_interop._cirq_to_bloq import _QReg
+
+    all_mapped_qubits = {q for qreg in qreg_to_qvar for q in qreg.qubits}
+    qubits_to_allocate = [q for q in in_reg.qubits if q not in all_mapped_qubits]
+    if qubits_to_allocate:
+        n_alloc = len(qubits_to_allocate)
+        qreg_to_qvar[
+            _QReg(qubits_to_allocate, dtype=qt.QBit() if n_alloc == 1 else qt.QAny(n_alloc))
+        ] = bb.allocate(n_alloc)
+
+    # if in_reg not in qreg_to_qvar: splits & joins needed, which shouldn't be the case
+    assert in_reg in qreg_to_qvar
+
+
+def _gather_input_soqs(bb: "qt.BloqBuilder", op_quregs, qreg_to_qvar):
+    """Modified function from Qualtran-Cirq interop module that collects input Soquets."""
+    qvars_in = {}
+    for reg_name, quregs in op_quregs.items():
+        flat_soqs: List[qt.Soquet] = []
+        for qureg in quregs.flatten():
+            _ensure_in_reg_exists(bb, qureg, qreg_to_qvar)
+            flat_soqs.append(qreg_to_qvar[qureg])
+        qvars_in[reg_name] = np.array(flat_soqs).reshape(quregs.shape)
+    return qvars_in
+
+
+def _inherit_from_bloq(cls):  # pylint: disable=too-many-statements
+    """Decorator for ToBloq to inherit from Qualtran's bloq only when qualtran is available."""
```

Comment:
Is there a more proper way to handle this? Are there any other similar examples in PL?

Reply:
I also couldnt find any similar examples of inheriting from an optional dependency's class in PL. When we work with optional dependencies, we do so functionally.